### PR TITLE
Optimized Java Versions

### PR DESCRIPTION
**Pull request by svanoort**
Includes content from https://github.com/juditacs/wordcount/pull/27 + fixes merge conflicts + adds a Java implementation using the Trove library, and storing Strings as UTF-8 encoded byte arrays. 

This reduces the memory footprint tremendously and allows the Java version to complete the full Wiki corpus without running out of memory, using a little bit over 4 GB of heap.   Performance relative to other implementations varies significantly depending on the configuration.

I'd be happy to remove excess Java implementations here (since there are many).  My thought would be to keep one "naive" Java version using a regex to split lines on tokens, and a HashMap<String,Integer> + sorting the Map.Entry methods, one of the optimized ones, and this version.

FYI @xupwup @sgwerder

### CODE REVIEW

**Comment by juditacs** [comment]
I already removed WordCountArray and on the long run, I plan to have only one version in each language. I'm open to compromise though, if you think it is worth comparing both Java versions, we can have two.

Let's wait for @xupwup's and  @sgwerder's comments before merging.

**Comment by sgwerder** [comment]
I'm fine with the Trove version. It runs slower than the previous ones for wikisource, but it completes the big file. Its a good new starting point for further little optimizations :) (Like checking out NIO and parallelization)
However it needs a cleanup and needs to pass the little tests.
Also consider this: [Sloppy Edit](http://pastebin.com/40niza0T)

**Comment by svanoort** [comment]
@sgwerder Yeah, it needs to be cleaned up and there's some room for more tweaking there (there are a few more tweaks in there).  Notes on a couple points before I call it a day on this:

> It runs slower than the previous ones for wikisource

It looks like there is a platform dependency here; on my linux box with OpenJDK7, I get the following:

Parsing/map addition time (ms): 5916
Creating Count objects for sorting
Count object creation time (ms): 90
sorting...
Sorting time (ms): 1587
output...
Output time (ms): 775

real	0m8.493s
user	0m11.396s
sys	0m0.464s

Vs. the other very optimized version:

real	0m9.083s
user	0m16.252s
sys	0m0.476s

On my newer Mac (Using Oracle JDK 8), they're both a bit faster, but the other version benefits even more, and comes out 20% faster.  For some reason sorting isn't any faster there.  Something to poke into definitely. 

> Also the command now runs with 6G max heap. Does hardware have 6G?

Yep, there's a reference to the Node.js test exceeding the 16 GB memory limit elsewhere, and some of the other implementations can spike that high on the full Wiki. 

> needs to pass the little tests 

Er, what are you seeing failing?  I am not seeing any failures on either system. 

> checking out NIO and parallelization

Spoiler: I've got a version that is halfway there to this, which is why I'm not worrying too much about cleanup yet - https://github.com/svanoort/wordcount/tree/svanoort-trove-allbinary

The biggest bottleneck for the Trove version so far is that it has to convert bytes to Unicode, then convert Unicode back to bytes to store it.  I tried using the NIO buffer methods to do this here - 7e19705718039aa46a0d39c84a78b68574a26b8a but it didn't prove to be any faster.  The experimental version above works directly with bytes and uses an NIO ByteBuffer for convenience. 

It could easily be adapted to read to a pool of buffers/maps and hand chunks of binary off to worker threads for tokenizing & counts.  The only problem is combining results at the end (probably a single-threaded operation).

I really miss Python after doing all this hacking around Java's limits.

**Comment by juditacs** [comment]
Btw I've been meaning to ask, why does the current Java implementation hang for hours instead of dying with memory error? JVM based languages (scala, clojure) at least die respectfully.

Let me know when your solution is merge-ready.

**Comment by sgwerder** [comment]
@svanoort This is getting fast now! I'm at 86 seconds usertime:

> Parsing and adding to map
Parsing/map addition time (ms): 58851
Creating Count objects for sorting
Count object creation time (ms): 541
sorting...
Sorting time (ms): 15187
output...
Output time (ms): 12344
WordCountTrove2: 86.932601244 seconds

If the test passes its ok. I guess its some weird Windows encoding issue on my end.

> I really miss Python after doing all this hacking around Java's limits.

Understandable. But you've been operating at the Java limits, Python has its own, similar frustrating ones. I remember getting furious one time when encoding, decoding just didn't work. Again probably some Windows issue. Then I heared of the GIL, but never needed multithreading myself.

> Btw I've been meaning to ask, why does the current Java implementation hang for hours instead of dying with memory error? JVM based languages (scala, clojure) at least die respectfully.

I got a "java.lang.OutOfMemoryError: Java heap space” or "java.lang.OutOfMemoryError: GC overhead limit exceeded" error after a long time (15 min). When I looked at the resource monitor I saw a typical high memory use with periodic dips. I guess when you give it so much more memory then it will try to carry on as long as possible with that. In the end it will be occupied with memory management.

File: java/WordCountTrove.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        System.err.println("Parsing/map addition time (ms): "+(endTime-startTime));
+
+        System.err.println("Creating Count objects for sorting");
+        startTime = System.currentTimeMillis();
+
+        final ArrayList<CountForWord> multiples = new ArrayList<CountForWord>(m.size()/2);
+        final ArrayList<byte[]> singles = new ArrayList<byte[]>(m.size()/2);
+
+        TObjectIntProcedure<byte[]> proc = new TObjectIntProcedure<byte[]>(){
+            int i=0;
```
Comment (sgwerder) [inline java/WordCountTrove.java]: what is this int i = 0; for?

Comment (svanoort) [inline java/WordCountTrove.java]: Heh, leftover junk from a previous implementation, oops!  (I was doing an array for performance reasons before, and needed an indexing variable).

**Comment by svanoort** [comment]
> I got a "java.lang.OutOfMemoryError: Java heap space” or "java.lang.OutOfMemoryError: GC overhead limit exceeded" error after a long time (15 min). When I looked at the resource monitor I saw a typical high memory use with periodic dips. I guess when you give it so much more memory then it will try to carry on as long as possible with that. In the end it will be occupied with memory management only until finally throwing an error.

This is correct.  More specifically, the default Java GC settings will only throw an OutOfMemoryError when spending 98% of time in GC and recovering less than 2% of the heap.  This is controlled by a tuneable parameter, and is only true for some GC mechanisms.

**Comment by sgwerder** [comment]
Recent changes made it slower again for my machine and for big file.

**Comment by svanoort** [comment]
@sgwerder Try the absolute latest version (it uses an all-bytes tokenizing pipeline), and use -Xmx6G -Xms4G  If's still slower, which step is slower?  Sorting?  It seems to be fast on both of my machines than WordCount, but there's room to tweak buffer sizes and the breakpoint at which we switch from builtin sort to radix sort. 

Is it looking cleaned up for you guys?  I think the optimized implementation is about as good as I can get it and fairly cleaned up.  I'd like to keep the logging of step information in there if it's okay, so it's possible to see where processing time goes. 

If juditacs is still up for it, I think it would make sense to have this + a baseline Java version (probably WordCount but with a vanilla StringTokenizer instead of the optimized character parsing).  That would reflect a smart implementation but without heavy optimization.  This would be a fully-customized version.

**Comment by sgwerder** [comment]
Your latest version is slower than your earlier version I commented on 10 hours ago. (With your suggested vm arguments):
Newest Version:

> Parsing and adding to map
Parsing/map addition time (ms): 70575
Creating Count objects for sorting
Count object creation time (ms): 962
sorting...
Sorting multiples time (ms): 9243 with count: 7189029
Sorting single time (ms): 19459 with count: 21612356
output...
Output time (ms): 6223
WordCount: 106.474797756 seconds

Earlier [version](http://pastebin.com/uP2PAzqL) (Sorry didnt find according commit):

> Parsing and adding to map
Parsing/map addition time (ms): 54851
Creating Count objects for sorting
Count object creation time (ms): 468
sorting...
Sorting time (ms): 15290
output...
Output time (ms): 6424
WordCount: 77.04467657 seconds

Looks like parsing and sorting went up, while output went down?

**Comment by svanoort** [comment]
@sgwerder Ahaha, the speed you see is actually due to a bug that causes it to abort tokenizing prematurely when it hits a token exceeding buffer size!  I discovered this when I noticed it was only returning 16M tokens rather than ~25M, but it was blazing fast - I fixed this in 0dfeb8ffe80d772ac7f3a55d705f79e51e0e8d60.  If you set the ByteBuffer to resize when a token exceeds it, you should see the performance drop but the number of token lines go up.  (The ByteBuffer version also returns incorrect token data due to some sort of complex off-by-one case). 

So, sorting is ~2x as fast because it's sorting roughly half as many elements, and parsing is faster because it's skipping much of the stream.   Output is still faster even with 2x as many tokens because now I directly dump bytes into the output stream (removing most of the Unicode overhead).

**Comment by svanoort** [comment]
Benchmarks on WikiSource:

**Trove optimal:**

Parsing and adding to map
Parsing/map addition time (ms): 2591
Creating Count objects for sorting
Count object creation time (ms): 99
sorting...
Sorting multiples time (ms): 521 with count: 593096
Sorting single time (ms): 941 with count: 1289781
output...
Output time (ms): 444

real    0m4.750s
user    0m5.635s
sys 0m0.449s


**Previous optimal without raw bytes/trove/radixsort:**
real    0m6.671s
user    0m7.624s
sys 0m0.685s


**Functional:**
real    0m12.308s
user    0m13.423s
sys 0m1.833s

# And on the full wiki:
**Trove optimized:** 
Runs happily in 4 GB of memory (and slightly faster with 6).  Time below with 4.

real    2m2.568s
user    2m50.963s
sys 0m5.635s

It'll actually complete with 2500 MB of RAM (no idea how, since that should be less than working set), but there's a lot of memory pressure and GC activity drives CPU time up to 8m20 & realtime to 2m:55s.  

**Functional version:**
Functional runs to 6 GB of memory, but really needs about 4.7 GB for working set, time below with 10 GB of heap but spiking to 6ish.

real    5m1.393s
user    7m2.312s
sys 0m27.953s

File: java/WordCountTrove.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+ *    - Buffer I/O, either using Buffered I/O classes or by explicitly reading chunks to an array
+ *    - Use Trove primitive collections (performance and memory savings, plus direct-iteration procedures)
+ *    - Process recurring tokens separately from singletons (big performance gain)
+ *    - Use a radix byte sort to sort large lists of bytestrings
+ *    - Sort strings using byte order (fudging the sort order slightly, for great performance savings)
+ *  
+ *  Authors:  
+ *   - Sam Van Oort <samvanoort@gmail.com>
+ *   - xupwup   
+ *   - Rick Hendricksen
```
Comment (xupwup) [inline java/WordCountTrove.java]: xupwup = Rick Hendricksen, by the way. That makes this kinda redundant.

Comment (svanoort) [inline java/WordCountTrove.java]: Aha! Duly noted, I also need to see about other credits here.

**Comment by svanoort** [comment]
@juditacs  I think we're ready to merge, assuming no objections from @xupwup or @sgwerder ?

I can think of a few possible ways to make the optimized version run *slightly* faster by tweaking the sorting algorithm or writing a custom HashMap implementation, but none that I know will yield significant speedups.

**Comment by juditacs** [comment]
Do you all agree that I can merge this?

**Comment by xupwup** [comment]
> Do you all agree that I can merge this?

I do

**Comment by sgwerder** [comment]
> Do you all agree that I can merge this?
Yes

**Comment by svanoort** [comment]
@juditacs Okay, let's merge! :smile: 

@xupwup Worth mentioning that I just stumbled across this: http://stackoverflow.com/questions/81346/most-efficient-way-to-increment-a-map-value-in-java/25354509#25354509 

Heh.  It's possible I might amend this solution *later* with an even faster custom hashtable version at some point, probably borrowing Koloboke's strategy here.  

One quick check revealed that if I compute (and use in a trivial way so it can't get JIT'd out) hashes, it's 17s to read, tokenize, create byte[], and hash it without storing to hashtable. Storing to hashtable makes this time ~65s.  If I preallocate the HashTable to be big enough for all tokens (10.5M with scrubbed data), it yields runtime of 54s, so 15% of the time is just table resizes (Trove does a rehash, so this is inefficient). 

I suspect I can do better than this with a highly custom hashtable implementation (that uses power-of-two table sizes and caches the hashes for resizing).  But coding and optimizing this is a few days work at least, so... a project for later.

**Comment by sgwerder** [comment]
I now just want it to beat the other languages and make them stare at this 303 LOC Java abomination :v: (not even counting Trove). This blog might be interesting for that HashTable project. (This and following entries): http://pzemtsov.github.io/2016/01/31/home-made-hash-tables-in-java.html

**Comment by svanoort** [comment]
@sgwerder It's a good goal, and a solid article; there's a few even better ones out there about the instruction cost of specific operations.

My goal is now to beat @gaebor and his current C implementation with Java by clever datastructure and sort algorithm use, despite Java's kind of terrible stream I/O.  At least on my box, I've got it running 2x as fast as the optimal python 2 version, so it's *possible* this will win in the next iteration. 

Sadly, this is not the most horrifying low-level Java I have written.  That crown belongs to a tightly optimized LZF implementation for the H2 database, or possibly the hand-tweaked loader for a complex binary format (I was trying to get them to load from disc fast enough to interact with in realtime).

**Comment by juditacs** [comment]
It's amazing! The trove version comes in 2nd place. It beats Cpp. First, I want to be absolutely that my measurement is correct.

Can you guys please remove the printouts to stderr?

**Comment by sgwerder** [comment]
@svanoort Just wanted to deposit this here for you as I run over it recently: http://stackoverflow.com/questions/81346/most-efficient-way-to-increment-a-map-value-in-java/25354509#25354509

From https://koloboke.com/ news (30.05.2016)

**Comment by svanoort** [comment]
@sgwerder So, trove is near-optimal for the insertion.  There's a way to make this even faster on insertion/checking though, by using a custom hashtable that exposes internals for use (allowing, for example checking of array equality by cached byte[] hashcode... and not creating a byte[] unless it's a net-new string). 

However micro-optimization of this is probably going to wait until I have time for pleasure coding.

**Pull request merged**